### ISSUE TITLE
web3j truffle generate & events & code reuse

### ISSUE THREAD

**Issue by urnilo** [web3j/web3j#1040]
# Events & code reuse

## _Issue_description_

I have the following two events in Solidity contracts that differ only in **name**:

```solidity
event NewSuperblock(bytes32 superblockHash, address who);
event ApprovedSuperblock(bytes32 superblockHash, address who);
```

and when I run `web3j truffle generate ...` on my contract, I can see in the generated Java file:

```java
   public static class NewSuperblockEventResponse {
        public Log log;
        public Bytes32 superblockHash;
        public Address who;
    }

    public static class ApprovedSuperblockEventResponse {
        public Log log;
        public Bytes32 superblockHash;
        public Address who;
    }
```

I would like my `NewSuperblockEventResponse` and `ApprovedSuperblockEventResponse` to implement a common interface so that I could work with these two events more easily in the rest of my Java code. So if the generated output were something like:

```java
   interface BaseResponse {
       Log getLog();
       Bytes32 getSuperblockHash();
       Address getWho();
   }

   public static class NewSuperblockEventResponse implement BaseResponse {
        public Log log;
        public Bytes32 superblockHash;
        public Address who;
 
        public Log getLog() { return log; }
        // and getter for superblockHash and who fields
    }

    public static class ApprovedSuperblockEventResponse {
        public Log log;
        public Bytes32 superblockHash;
        public Address who;

        public Log getLog() { return log; }
        // and getter for superblockHash and who fields
    }
```

it would be really great.

What do you think? Is there any other way to achieve what I'm trying to do? Or am I just doing what I'm not supposed to do in the first place?

Thanks

**Comment by iikirilov** [comment]
It is possible - but it is quite complicated to generate this. You would have to track all the events and then generate a base class for events with common methods. The issue is how would you handle a contract with many events where some events have some common methods and other events have other common methods so you would have multiple base classes. How would you name those base classes?

We have already merged #970 which is a step in the direction you want. Right now I would do this manually - just change the generated code as you have suggested.

Another option you have is to to have a condition like `if (e instanceof MyEvent || e instanceof MyOtherEvent)` and then cast/handle that event type appropriately.

Another option is reflection but that is not a good approach because if you change the event things will break.

@xaviarias do you have an opinion on this.

**Comment by iikirilov** [comment]
I think the cost of doing this is far greater than the benefit considering you can do this manually. I know it is a manual step but you could automate this with a gradle task if you need to.
